### PR TITLE
Adding, renaming, fixing JwtBuilder methods

### PR DESCRIPTION
**Pull request by abatishchev**

### CODE REVIEW

File: src/JWT/Algorithms/HMACSHAAlgorithmFactory.cs [hunk 1]
```diff
@@ -25,7 +25,7 @@ public virtual IJwtAlgorithm Create(JwtHashAlgorithm algorithm)
                 case JwtHashAlgorithm.RS256:
                     throw new NotSupportedException($"For algorithm {nameof(JwtHashAlgorithm.RS256)} please create custom factory by implementing {nameof(IAlgorithmFactory)}");
                 default:
-                    throw new InvalidOperationException($"Algorithm {algorithm} is not supported.");
+                    throw new NotSupportedException($"Algorithm {algorithm} is not supported.");
```
Comment (nbarbettini) [inline src/JWT/Algorithms/HMACSHAAlgorithmFactory.cs]: Technically a breaking change. Will this get a version bump?

File: src/JWT/Builder/JwtBuilder.cs [hunk 2]
```diff
@@ -164,101 +187,126 @@ public JwtBuilder SetVerifySignature(bool verify)
...
+                throw new InvalidOperationException("Can't decode a token. Check if you have call all of the followng methods:\r\n" +
+                                                    $"-{nameof(SetSerializer)}\r\n" +
+                                                    $"-{nameof(SetValidator)}\r\n" +
+                                                    $"-{nameof(SetUrlEncoder)}\r\n" +
+                                                    $"If you called {nameof(MustVerifySignature)} you must also call {nameof(SetSecret)}.");
         }
 
         /// <summary>
-        /// Checks whether enought information was supplied to build a new token.
+        /// Checks whether enought dependencies were supplied to build a new token.
```
Comment (nbarbettini) [inline src/JWT/Builder/JwtBuilder.cs]: nitpick: `enough`

File: src/JWT/Builder/JwtBuilder.cs [hunk 3]
```diff
@@ -164,101 +187,126 @@ public JwtBuilder SetVerifySignature(bool verify)
...
                    _urlEncoder != null &&
                    _jwt.Payload != null &&
-                   _secret != null &&
-                   _secret.Length > 0;
+                   !String.IsNullOrEmpty(_secret);
         }
 
         /// <summary>
-        /// Checks whether enought information was supplied to decode a token.
+        /// Checks whether enought dependencies were supplied to decode a token.
```
Comment (nbarbettini) [inline src/JWT/Builder/JwtBuilder.cs]: `enough`

File: src/JWT/Algorithms/HMACSHAAlgorithmFactory.cs (hunk 1 above)
Comment (abatishchev) [reply src/JWT/Algorithms/HMACSHAAlgorithmFactory.cs]: Yes, this and few minor more. So wondering too should we bump to 4.0 or keep 3.2

**Comment by abatishchev** [comment]
@nbarbettini @paule96 another question whether we should rename all builder's method to `With<Bla>()` for consistency across the project and other projects' fluent apis.

**Comment by nbarbettini** [comment]
@abatishchev Like actually using generics, or just `WithIssuer`, `WithAudience`, etc?

**Comment by abatishchev** [comment]
Just `WithIssuer`, `WithAudience`, etc. I think it's more common along similar fluent builders.

**Comment by nbarbettini** [comment]
Yeah, I agree that it's more common. I don't have a strong opinion either way; `WithX` is fine with me.

**Comment by abatishchev** [comment]
Renamed all method to With*

**Comment by abatishchev** [comment]
Pushed to NuGet as [4.0.0-beta1](https://www.nuget.org/packages/JWT/4.0.0-beta1).

**Pull request merged**